### PR TITLE
[bench]: state-tree get_with_proof and commit api

### PR DESCRIPTION
**Pull request by lerencao**
`cargo bench -p starcoin-state-tree -- put_and_commit`:

```
put_and_commit/mem_store/1
                        time:   [120.50 us 121.89 us 123.31 us]
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe
put_and_commit/db_store/1
                        time:   [120.28 us 121.59 us 122.97 us]
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe
put_and_commit/mem_store/5
                        time:   [562.98 us 570.24 us 577.69 us]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
put_and_commit/db_store/5
                        time:   [563.89 us 573.33 us 582.83 us]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
put_and_commit/mem_store/10
                        time:   [1.1257 ms 1.1420 ms 1.1582 ms]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
put_and_commit/db_store/10
                        time:   [1.1279 ms 1.1458 ms 1.1645 ms]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
Benchmarking put_and_commit/mem_store/50: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 26.4s or reduce sample count to 30.
put_and_commit/mem_store/50
                        time:   [5.9393 ms 6.0492 ms 6.1613 ms]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
Benchmarking put_and_commit/db_store/50: Warming up for 3.0000 s

Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 25.1s or reduce sample count to 30.
put_and_commit/db_store/50
                        time:   [5.9074 ms 5.9889 ms 6.0716 ms]
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe
Benchmarking put_and_commit/mem_store/100: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 46.3s or reduce sample count to 20.
Benchmarking put_and_commit/mem_store/100: Collecting 100 samples in estimated 46.340 s (5050 iterations)
put_and_commit/mem_store/100
                        time:   [12.487 ms 12.712 ms 12.979 ms]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
Benchmarking put_and_commit/db_store/100: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 46.9s or reduce sample count to 20.
Benchmarking put_and_commit/db_store/100: Collecting 100 samples in estimated 46.897 s (5050 iterations)

put_and_commit/db_store/100
                        time:   [13.220 ms 13.638 ms 14.061 ms]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```

### CODE REVIEW

**Comment by lerencao** [comment]
get_with_proof:

```
get_with_proof/mem_store
                        time:   [41.105 us 42.053 us 43.192 us]
                        change: [-68.716% -58.606% -46.699%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe
get_with_proof/db_store time:   [62.505 us 63.731 us 65.225 us]
                        change: [-80.333% -63.727% -30.503%] (p = 0.04 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe
```

**Comment by lerencao** [comment]
The result bench runs on MBP 8G, 2Core.

File: state/state-tree/src/mock/mod.rs [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
+use starcoin_state_store_api::{StateNode, StateNodeStore};
 
 use starcoin_crypto::HashValue;
 use std::collections::{BTreeMap, HashMap};
-use std::sync::Mutex;
-
+use std::sync::RwLock;
 #[derive(Default)]
 pub struct MockStateNodeStore {
-    nodes: Mutex<HashMap<HashValue, StateNode>>,
```
Comment (ssyuan) [inline state/state-tree/src/mock/mod.rs:13]: 我原来加node的rpc接口时，编译不过，必须得用mutex，你确认下

File: benchmarks/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,18 @@ storage = { package="starcoin-storage", path = "../storage"}
 crypto = { package="starcoin-crypto", path = "../commons/crypto"}
 starcoin-config = { path = "../config"}
 starcoin-accumulator = { package="starcoin-accumulator", path = "../core/accumulator"}
+forkable-jellyfish-merkle = { path = "../core/forkable-jellyfish-merkle"}
+starcoin-state-store-api = {path = "../state/state-store-api"}
+starcoin-state-tree = {path = "../state/state-tree"}
+
+[dev-dependencies]
+rand = "0.5"
+tempfile = "3"
```
[the remark this answers is not in the document]
Comment (lerencao) [reply benchmarks/Cargo.toml]: Updated.

File: state/state-tree/src/mock/mod.rs (hunk 1 above)
Comment (lerencao) [reply state/state-tree/src/mock/mod.rs:13]: CI is passed.

**Comment by ssyuan** [comment]
50k result:

get_with_proof/mem_store                                                                             
                        time:   [30.888 us 31.108 us 31.315 us]
                        change: [-8.1249% -7.0129% -5.8420%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
get_with_proof/db_store time:   [48.860 us 49.657 us 50.461 us]                                     
                        change: [-64.047% -58.126% -50.517%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

put_and_commit/mem_store/1                                                                            
                        time:   [222.66 us 259.74 us 299.09 us]
                        change: [-10.589% +11.034% +37.304%] (p = 0.36 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  10 (10.00%) high mild
  1 (1.00%) high severe
put_and_commit/db_store/1                                                                            
                        time:   [141.05 us 150.25 us 162.51 us]
                        change: [+94.672% +119.89% +147.26%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
put_and_commit/mem_store/5                                                                            
                        time:   [518.96 us 524.61 us 530.37 us]
                        change: [+7.9065% +9.2870% +10.766%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
put_and_commit/db_store/5                                                                            
                        time:   [531.39 us 536.42 us 541.48 us]
                        change: [-68.015% -62.040% -54.201%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
Benchmarking put_and_commit/mem_store/10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.1s or reduce sample count to 60.
put_and_commit/mem_store/10                                                                             
                        time:   [997.00 us 1.0053 ms 1.0149 ms]
                        change: [+2.8678% +4.6887% +6.4266%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
Benchmarking put_and_commit/db_store/10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.5s or reduce sample count to 60.
put_and_commit/db_store/10                                                                             
                        time:   [1.0699 ms 1.0871 ms 1.1080 ms]
                        change: [+14.318% +15.978% +17.724%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
Benchmarking put_and_commit/mem_store/50: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 25.4s or reduce sample count to 30.
put_and_commit/mem_store/50                                                                             
                        time:   [5.2684 ms 5.3478 ms 5.4260 ms]
                        change: [+9.0861% +10.470% +11.963%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  7 (7.00%) high mild
  1 (1.00%) high severe
Benchmarking put_and_commit/db_store/50: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 26.5s or reduce sample count to 30.
put_and_commit/db_store/50                                                                             
                        time:   [5.8250 ms 6.3588 ms 7.0315 ms]
                        change: [-25.347% -9.6829% +9.0261%] (p = 0.31 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) high mild
  10 (10.00%) high severe
Benchmarking put_and_commit/mem_store/100: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 51.1s or reduce sample count to 20.
put_and_commit/mem_store/100                                                                             
                        time:   [11.533 ms 12.341 ms 13.395 ms]
                        change: [+18.707% +29.388% +43.146%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe
Benchmarking put_and_commit/db_store/100: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 54.0s or reduce sample count to 20.
put_and_commit/db_store/100                                                                             
                        time:   [11.498 ms 12.300 ms 13.355 ms]
                        change: [-1.0283% +11.352% +23.623%] (p = 0.05 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)

**Pull request merged**